src/System.Private.CoreLib/shared/System/BitOps.cs

```diff
@@ -49,15 +57,183 @@ public static int TrailingZeroCount(uint value)
                 return (int)Bmi1.TrailingZeroCount(value);
             }
 
-            // Main code has behavior 0->0, so special-case in order to match intrinsic path 0->32
-            if (value == 0u)
+            // Software fallback has behavior 0->0, so special-case to match intrinsic path 0->32
+            if (value == 0)
+            {
                 return 32;
+            }
 
             // uint.MaxValue >> 27 is always in range [0 - 31] so we use Unsafe.AddByteOffset to avoid bounds check
             return Unsafe.AddByteOffset(
-                ref MemoryMarshal.GetReference(s_TrailingZeroCountDeBruijn), 
                 // Using deBruijn sequence, k=2, n=5 (2^5=32) : 0b_0000_0111_0111_1100_1011_0101_0011_0001u
-                ((uint)((value & -value) * 0x077CB531u)) >> 27);
+                ref MemoryMarshal.GetReference(s_TrailingZeroCountDeBruijn),
+                // long -> IntPtr cast on 32-bit platforms is expensive - it does overflow checks not needed here
+                (IntPtr)(int)(((uint)((value & -value) * 0x077CB531u)) >> 27)); // shift over long also expensive on 32-bit
```

Comment:
Couldn't this be made significantly simpler by just inserting the cast in front of `-value`. That is: `((value & (uint)(-value)) * 0x077CB531u) >> 27`

Reply:
The problem, as I see it, is that `-value` where `uint` is a value returns a `long`. This also means we will be doing 64-bit multiplication and a 64-bit shift, both of which look unnecessary.

Reply:
@tannergooding do you mind if I look at this in the next PR (which is waiting to go), I would like to get this one in?

Reply:
OK, fixed in subsequent PR (and units still pass), thanks - that is much easier to read